{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "214054ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install gseapy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d189a185",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from gseapy.plot import gseaplot\n",
    "import gseapy as gp\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "272eccc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>320.481168</td>\n",
       "      <td>-0.492442</td>\n",
       "      <td>0.165330</td>\n",
       "      <td>-2.978533</td>\n",
       "      <td>0.002896</td>\n",
       "      <td>0.036899</td>\n",
       "      <td>ATAD3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000069424</td>\n",
       "      <td>231.438079</td>\n",
       "      <td>1.135118</td>\n",
       "      <td>0.274328</td>\n",
       "      <td>4.137818</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.001414</td>\n",
       "      <td>KCNAB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSG00000116786</td>\n",
       "      <td>1216.590493</td>\n",
       "      <td>0.390974</td>\n",
       "      <td>0.122426</td>\n",
       "      <td>3.193548</td>\n",
       "      <td>0.001405</td>\n",
       "      <td>0.022625</td>\n",
       "      <td>PLEKHM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>ENSG00000204138</td>\n",
       "      <td>315.739587</td>\n",
       "      <td>-0.541585</td>\n",
       "      <td>0.154028</td>\n",
       "      <td>-3.516146</td>\n",
       "      <td>0.000438</td>\n",
       "      <td>0.009790</td>\n",
       "      <td>PHACTR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>ENSG00000188529</td>\n",
       "      <td>1167.144954</td>\n",
       "      <td>-0.356988</td>\n",
       "      <td>0.123730</td>\n",
       "      <td>-2.885228</td>\n",
       "      <td>0.003911</td>\n",
       "      <td>0.045829</td>\n",
       "      <td>SRSF10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10824</th>\n",
       "      <td>ENSG00000159131</td>\n",
       "      <td>999.681860</td>\n",
       "      <td>-0.383878</td>\n",
       "      <td>0.134314</td>\n",
       "      <td>-2.858067</td>\n",
       "      <td>0.004262</td>\n",
       "      <td>0.048859</td>\n",
       "      <td>GART</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10826</th>\n",
       "      <td>ENSG00000159259</td>\n",
       "      <td>114.852294</td>\n",
       "      <td>-1.795568</td>\n",
       "      <td>0.463159</td>\n",
       "      <td>-3.876783</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>0.003297</td>\n",
       "      <td>CHAF1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10863</th>\n",
       "      <td>ENSG00000160298</td>\n",
       "      <td>76.683762</td>\n",
       "      <td>-1.196280</td>\n",
       "      <td>0.363275</td>\n",
       "      <td>-3.293041</td>\n",
       "      <td>0.000991</td>\n",
       "      <td>0.017533</td>\n",
       "      <td>C21orf58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10878</th>\n",
       "      <td>ENSG00000159055</td>\n",
       "      <td>123.457054</td>\n",
       "      <td>-0.951679</td>\n",
       "      <td>0.318707</td>\n",
       "      <td>-2.986066</td>\n",
       "      <td>0.002826</td>\n",
       "      <td>0.036395</td>\n",
       "      <td>MIS18A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10909</th>\n",
       "      <td>ENSG00000159200</td>\n",
       "      <td>1180.226926</td>\n",
       "      <td>1.546800</td>\n",
       "      <td>0.472981</td>\n",
       "      <td>3.270323</td>\n",
       "      <td>0.001074</td>\n",
       "      <td>0.018582</td>\n",
       "      <td>RCAN1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1189 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0     baseMean  log2FoldChange     lfcSE      stat  \\\n",
       "0      ENSG00000160072   320.481168       -0.492442  0.165330 -2.978533   \n",
       "19     ENSG00000069424   231.438079        1.135118  0.274328  4.137818   \n",
       "21     ENSG00000116786  1216.590493        0.390974  0.122426  3.193548   \n",
       "65     ENSG00000204138   315.739587       -0.541585  0.154028 -3.516146   \n",
       "77     ENSG00000188529  1167.144954       -0.356988  0.123730 -2.885228   \n",
       "...                ...          ...             ...       ...       ...   \n",
       "10824  ENSG00000159131   999.681860       -0.383878  0.134314 -2.858067   \n",
       "10826  ENSG00000159259   114.852294       -1.795568  0.463159 -3.876783   \n",
       "10863  ENSG00000160298    76.683762       -1.196280  0.363275 -3.293041   \n",
       "10878  ENSG00000159055   123.457054       -0.951679  0.318707 -2.986066   \n",
       "10909  ENSG00000159200  1180.226926        1.546800  0.472981  3.270323   \n",
       "\n",
       "         pvalue      padj      Gene  \n",
       "0      0.002896  0.036899    ATAD3B  \n",
       "19     0.000035  0.001414    KCNAB2  \n",
       "21     0.001405  0.022625   PLEKHM2  \n",
       "65     0.000438  0.009790   PHACTR4  \n",
       "77     0.003911  0.045829    SRSF10  \n",
       "...         ...       ...       ...  \n",
       "10824  0.004262  0.048859      GART  \n",
       "10826  0.000106  0.003297    CHAF1B  \n",
       "10863  0.000991  0.017533  C21orf58  \n",
       "10878  0.002826  0.036395    MIS18A  \n",
       "10909  0.001074  0.018582     RCAN1  \n",
       "\n",
       "[1189 rows x 8 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('DE_results.csv').dropna()\n",
    "df = df.rename(columns = {'symbol': 'Gene'})\n",
    "\n",
    "# it is best practice not to filter?\n",
    "df = df[df.padj < 0.05].dropna()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ea48c308",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000261115</td>\n",
       "      <td>193.721808</td>\n",
       "      <td>4.741075</td>\n",
       "      <td>0.537213</td>\n",
       "      <td>8.825318</td>\n",
       "      <td>1.091489e-18</td>\n",
       "      <td>1.672817e-14</td>\n",
       "      <td>TMEM178B</td>\n",
       "      <td>65.315669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000145390</td>\n",
       "      <td>2016.886474</td>\n",
       "      <td>2.249586</td>\n",
       "      <td>0.263078</td>\n",
       "      <td>8.551018</td>\n",
       "      <td>1.220076e-17</td>\n",
       "      <td>6.232962e-14</td>\n",
       "      <td>USP53</td>\n",
       "      <td>29.706469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000109846</td>\n",
       "      <td>524.720804</td>\n",
       "      <td>2.398381</td>\n",
       "      <td>0.348668</td>\n",
       "      <td>6.878701</td>\n",
       "      <td>6.040066e-12</td>\n",
       "      <td>9.257006e-09</td>\n",
       "      <td>CRYAB</td>\n",
       "      <td>19.267466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000106483</td>\n",
       "      <td>7321.978396</td>\n",
       "      <td>3.438204</td>\n",
       "      <td>0.602508</td>\n",
       "      <td>5.706491</td>\n",
       "      <td>1.153292e-08</td>\n",
       "      <td>3.668965e-06</td>\n",
       "      <td>SFRP4</td>\n",
       "      <td>18.688209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000277586</td>\n",
       "      <td>78.098744</td>\n",
       "      <td>3.388113</td>\n",
       "      <td>0.677038</td>\n",
       "      <td>5.004316</td>\n",
       "      <td>5.606066e-07</td>\n",
       "      <td>6.411834e-05</td>\n",
       "      <td>NEFL</td>\n",
       "      <td>14.206417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1184</th>\n",
       "      <td>ENSG00000203760</td>\n",
       "      <td>107.203668</td>\n",
       "      <td>-1.974640</td>\n",
       "      <td>0.272572</td>\n",
       "      <td>-7.244486</td>\n",
       "      <td>4.340832e-13</td>\n",
       "      <td>1.108793e-09</td>\n",
       "      <td>CENPW</td>\n",
       "      <td>-17.683201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1185</th>\n",
       "      <td>ENSG00000170011</td>\n",
       "      <td>53.464280</td>\n",
       "      <td>-4.163618</td>\n",
       "      <td>0.803520</td>\n",
       "      <td>-5.181720</td>\n",
       "      <td>2.198487e-07</td>\n",
       "      <td>3.311684e-05</td>\n",
       "      <td>MYRIP</td>\n",
       "      <td>-18.652804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1186</th>\n",
       "      <td>ENSG00000176208</td>\n",
       "      <td>75.279706</td>\n",
       "      <td>-2.164796</td>\n",
       "      <td>0.304987</td>\n",
       "      <td>-7.097984</td>\n",
       "      <td>1.265898e-12</td>\n",
       "      <td>2.155684e-09</td>\n",
       "      <td>ATAD5</td>\n",
       "      <td>-18.761020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1187</th>\n",
       "      <td>ENSG00000024526</td>\n",
       "      <td>341.005688</td>\n",
       "      <td>-2.632820</td>\n",
       "      <td>0.386997</td>\n",
       "      <td>-6.803209</td>\n",
       "      <td>1.023138e-11</td>\n",
       "      <td>1.425510e-08</td>\n",
       "      <td>DEPDC1</td>\n",
       "      <td>-20.657185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1188</th>\n",
       "      <td>ENSG00000211772</td>\n",
       "      <td>80.031143</td>\n",
       "      <td>-6.688654</td>\n",
       "      <td>1.082359</td>\n",
       "      <td>-6.179698</td>\n",
       "      <td>6.422429e-10</td>\n",
       "      <td>3.785775e-07</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>-42.953503</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1189 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Unnamed: 0     baseMean  log2FoldChange     lfcSE      stat  \\\n",
       "0     ENSG00000261115   193.721808        4.741075  0.537213  8.825318   \n",
       "1     ENSG00000145390  2016.886474        2.249586  0.263078  8.551018   \n",
       "2     ENSG00000109846   524.720804        2.398381  0.348668  6.878701   \n",
       "3     ENSG00000106483  7321.978396        3.438204  0.602508  5.706491   \n",
       "4     ENSG00000277586    78.098744        3.388113  0.677038  5.004316   \n",
       "...               ...          ...             ...       ...       ...   \n",
       "1184  ENSG00000203760   107.203668       -1.974640  0.272572 -7.244486   \n",
       "1185  ENSG00000170011    53.464280       -4.163618  0.803520 -5.181720   \n",
       "1186  ENSG00000176208    75.279706       -2.164796  0.304987 -7.097984   \n",
       "1187  ENSG00000024526   341.005688       -2.632820  0.386997 -6.803209   \n",
       "1188  ENSG00000211772    80.031143       -6.688654  1.082359 -6.179698   \n",
       "\n",
       "            pvalue          padj      Gene       Rank  \n",
       "0     1.091489e-18  1.672817e-14  TMEM178B  65.315669  \n",
       "1     1.220076e-17  6.232962e-14     USP53  29.706469  \n",
       "2     6.040066e-12  9.257006e-09     CRYAB  19.267466  \n",
       "3     1.153292e-08  3.668965e-06     SFRP4  18.688209  \n",
       "4     5.606066e-07  6.411834e-05      NEFL  14.206417  \n",
       "...            ...           ...       ...        ...  \n",
       "1184  4.340832e-13  1.108793e-09     CENPW -17.683201  \n",
       "1185  2.198487e-07  3.311684e-05     MYRIP -18.652804  \n",
       "1186  1.265898e-12  2.155684e-09     ATAD5 -18.761020  \n",
       "1187  1.023138e-11  1.425510e-08    DEPDC1 -20.657185  \n",
       "1188  6.422429e-10  3.785775e-07     TRBC2 -42.953503  \n",
       "\n",
       "[1189 rows x 9 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rank sample with p-value and fold change. Rank the gene from positively enriched to negatively enriched\n",
    "df['Rank'] = -np.log10(df.padj)*df.log2FoldChange\n",
    "\n",
    "df = df.sort_values('Rank', ascending = False).reset_index(drop = True)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9d10aa29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TMEM178B</td>\n",
       "      <td>65.315669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>USP53</td>\n",
       "      <td>29.706469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CRYAB</td>\n",
       "      <td>19.267466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SFRP4</td>\n",
       "      <td>18.688209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NEFL</td>\n",
       "      <td>14.206417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1184</th>\n",
       "      <td>CENPW</td>\n",
       "      <td>-17.683201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1185</th>\n",
       "      <td>MYRIP</td>\n",
       "      <td>-18.652804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1186</th>\n",
       "      <td>ATAD5</td>\n",
       "      <td>-18.761020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1187</th>\n",
       "      <td>DEPDC1</td>\n",
       "      <td>-20.657185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1188</th>\n",
       "      <td>TRBC2</td>\n",
       "      <td>-42.953503</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1189 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene       Rank\n",
       "0     TMEM178B  65.315669\n",
       "1        USP53  29.706469\n",
       "2        CRYAB  19.267466\n",
       "3        SFRP4  18.688209\n",
       "4         NEFL  14.206417\n",
       "...        ...        ...\n",
       "1184     CENPW -17.683201\n",
       "1185     MYRIP -18.652804\n",
       "1186     ATAD5 -18.761020\n",
       "1187    DEPDC1 -20.657185\n",
       "1188     TRBC2 -42.953503\n",
       "\n",
       "[1189 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ranking = df[['Gene', 'Rank']]\n",
    "ranking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2b2241a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ARCHS4_Cell-lines',\n",
       " 'ARCHS4_IDG_Coexp',\n",
       " 'ARCHS4_Kinases_Coexp',\n",
       " 'ARCHS4_TFs_Coexp',\n",
       " 'ARCHS4_Tissues',\n",
       " 'Achilles_fitness_decrease',\n",
       " 'Achilles_fitness_increase',\n",
       " 'Aging_Perturbations_from_GEO_down',\n",
       " 'Aging_Perturbations_from_GEO_up',\n",
       " 'Allen_Brain_Atlas_10x_scRNA_2021',\n",
       " 'Allen_Brain_Atlas_down',\n",
       " 'Allen_Brain_Atlas_up',\n",
       " 'Azimuth_Cell_Types_2021',\n",
       " 'BioCarta_2013',\n",
       " 'BioCarta_2015',\n",
       " 'BioCarta_2016',\n",
       " 'BioPlanet_2019',\n",
       " 'BioPlex_2017',\n",
       " 'CCLE_Proteomics_2020',\n",
       " 'CORUM',\n",
       " 'COVID-19_Related_Gene_Sets',\n",
       " 'COVID-19_Related_Gene_Sets_2021',\n",
       " 'Cancer_Cell_Line_Encyclopedia',\n",
       " 'CellMarker_Augmented_2021',\n",
       " 'ChEA_2013',\n",
       " 'ChEA_2015',\n",
       " 'ChEA_2016',\n",
       " 'ChEA_2022',\n",
       " 'Chromosome_Location',\n",
       " 'Chromosome_Location_hg19',\n",
       " 'ClinVar_2019',\n",
       " 'DSigDB',\n",
       " 'Data_Acquisition_Method_Most_Popular_Genes',\n",
       " 'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
       " 'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
       " 'Descartes_Cell_Types_and_Tissue_2021',\n",
       " 'Diabetes_Perturbations_GEO_2022',\n",
       " 'DisGeNET',\n",
       " 'Disease_Perturbations_from_GEO_down',\n",
       " 'Disease_Perturbations_from_GEO_up',\n",
       " 'Disease_Signatures_from_GEO_down_2014',\n",
       " 'Disease_Signatures_from_GEO_up_2014',\n",
       " 'DrugMatrix',\n",
       " 'Drug_Perturbations_from_GEO_2014',\n",
       " 'Drug_Perturbations_from_GEO_down',\n",
       " 'Drug_Perturbations_from_GEO_up',\n",
       " 'ENCODE_Histone_Modifications_2013',\n",
       " 'ENCODE_Histone_Modifications_2015',\n",
       " 'ENCODE_TF_ChIP-seq_2014',\n",
       " 'ENCODE_TF_ChIP-seq_2015',\n",
       " 'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
       " 'ESCAPE',\n",
       " 'Elsevier_Pathway_Collection',\n",
       " 'Enrichr_Libraries_Most_Popular_Genes',\n",
       " 'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
       " 'Enrichr_Users_Contributed_Lists_2020',\n",
       " 'Epigenomics_Roadmap_HM_ChIP-seq',\n",
       " 'FANTOM6_lncRNA_KD_DEGs',\n",
       " 'GO_Biological_Process_2013',\n",
       " 'GO_Biological_Process_2015',\n",
       " 'GO_Biological_Process_2017',\n",
       " 'GO_Biological_Process_2017b',\n",
       " 'GO_Biological_Process_2018',\n",
       " 'GO_Biological_Process_2021',\n",
       " 'GO_Biological_Process_2023',\n",
       " 'GO_Cellular_Component_2013',\n",
       " 'GO_Cellular_Component_2015',\n",
       " 'GO_Cellular_Component_2017',\n",
       " 'GO_Cellular_Component_2017b',\n",
       " 'GO_Cellular_Component_2018',\n",
       " 'GO_Cellular_Component_2021',\n",
       " 'GO_Cellular_Component_2023',\n",
       " 'GO_Molecular_Function_2013',\n",
       " 'GO_Molecular_Function_2015',\n",
       " 'GO_Molecular_Function_2017',\n",
       " 'GO_Molecular_Function_2017b',\n",
       " 'GO_Molecular_Function_2018',\n",
       " 'GO_Molecular_Function_2021',\n",
       " 'GO_Molecular_Function_2023',\n",
       " 'GTEx_Aging_Signatures_2021',\n",
       " 'GTEx_Tissue_Expression_Down',\n",
       " 'GTEx_Tissue_Expression_Up',\n",
       " 'GTEx_Tissues_V8_2023',\n",
       " 'GWAS_Catalog_2019',\n",
       " 'GWAS_Catalog_2023',\n",
       " 'GeDiPNet_2023',\n",
       " 'GeneSigDB',\n",
       " 'Gene_Perturbations_from_GEO_down',\n",
       " 'Gene_Perturbations_from_GEO_up',\n",
       " 'Genes_Associated_with_NIH_Grants',\n",
       " 'Genome_Browser_PWMs',\n",
       " 'GlyGen_Glycosylated_Proteins_2022',\n",
       " 'HDSigDB_Human_2021',\n",
       " 'HDSigDB_Mouse_2021',\n",
       " 'HMDB_Metabolites',\n",
       " 'HMS_LINCS_KinomeScan',\n",
       " 'HomoloGene',\n",
       " 'HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression',\n",
       " 'HuBMAP_ASCTplusB_augmented_2022',\n",
       " 'HumanCyc_2015',\n",
       " 'HumanCyc_2016',\n",
       " 'Human_Gene_Atlas',\n",
       " 'Human_Phenotype_Ontology',\n",
       " 'IDG_Drug_Targets_2022',\n",
       " 'InterPro_Domains_2019',\n",
       " 'Jensen_COMPARTMENTS',\n",
       " 'Jensen_DISEASES',\n",
       " 'Jensen_TISSUES',\n",
       " 'KEA_2013',\n",
       " 'KEA_2015',\n",
       " 'KEGG_2013',\n",
       " 'KEGG_2015',\n",
       " 'KEGG_2016',\n",
       " 'KEGG_2019_Human',\n",
       " 'KEGG_2019_Mouse',\n",
       " 'KEGG_2021_Human',\n",
       " 'KOMP2_Mouse_Phenotypes_2022',\n",
       " 'Kinase_Perturbations_from_GEO_down',\n",
       " 'Kinase_Perturbations_from_GEO_up',\n",
       " 'L1000_Kinase_and_GPCR_Perturbations_down',\n",
       " 'L1000_Kinase_and_GPCR_Perturbations_up',\n",
       " 'LINCS_L1000_CRISPR_KO_Consensus_Sigs',\n",
       " 'LINCS_L1000_Chem_Pert_Consensus_Sigs',\n",
       " 'LINCS_L1000_Chem_Pert_down',\n",
       " 'LINCS_L1000_Chem_Pert_up',\n",
       " 'LINCS_L1000_Ligand_Perturbations_down',\n",
       " 'LINCS_L1000_Ligand_Perturbations_up',\n",
       " 'Ligand_Perturbations_from_GEO_down',\n",
       " 'Ligand_Perturbations_from_GEO_up',\n",
       " 'MAGMA_Drugs_and_Diseases',\n",
       " 'MAGNET_2023',\n",
       " 'MCF7_Perturbations_from_GEO_down',\n",
       " 'MCF7_Perturbations_from_GEO_up',\n",
       " 'MGI_Mammalian_Phenotype_2013',\n",
       " 'MGI_Mammalian_Phenotype_2017',\n",
       " 'MGI_Mammalian_Phenotype_Level_3',\n",
       " 'MGI_Mammalian_Phenotype_Level_4',\n",
       " 'MGI_Mammalian_Phenotype_Level_4_2019',\n",
       " 'MGI_Mammalian_Phenotype_Level_4_2021',\n",
       " 'MSigDB_Computational',\n",
       " 'MSigDB_Hallmark_2020',\n",
       " 'MSigDB_Oncogenic_Signatures',\n",
       " 'Metabolomics_Workbench_Metabolites_2022',\n",
       " 'Microbe_Perturbations_from_GEO_down',\n",
       " 'Microbe_Perturbations_from_GEO_up',\n",
       " 'Mouse_Gene_Atlas',\n",
       " 'NCI-60_Cancer_Cell_Lines',\n",
       " 'NCI-Nature_2015',\n",
       " 'NCI-Nature_2016',\n",
       " 'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
       " 'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
       " 'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
       " 'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
       " 'NURSA_Human_Endogenous_Complexome',\n",
       " 'OMIM_Disease',\n",
       " 'OMIM_Expanded',\n",
       " 'Old_CMAP_down',\n",
       " 'Old_CMAP_up',\n",
       " 'Orphanet_Augmented_2021',\n",
       " 'PFOCR_Pathways',\n",
       " 'PFOCR_Pathways_2023',\n",
       " 'PPI_Hub_Proteins',\n",
       " 'PanglaoDB_Augmented_2021',\n",
       " 'Panther_2015',\n",
       " 'Panther_2016',\n",
       " 'Pfam_Domains_2019',\n",
       " 'Pfam_InterPro_Domains',\n",
       " 'PheWeb_2019',\n",
       " 'PhenGenI_Association_2021',\n",
       " 'Phosphatase_Substrates_from_DEPOD',\n",
       " 'ProteomicsDB_2020',\n",
       " 'Proteomics_Drug_Atlas_2023',\n",
       " 'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Human_Down',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Human_Up',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Mouse_Down',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Mouse_Up',\n",
       " 'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
       " 'Rare_Diseases_AutoRIF_Gene_Lists',\n",
       " 'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
       " 'Rare_Diseases_GeneRIF_Gene_Lists',\n",
       " 'Reactome_2013',\n",
       " 'Reactome_2015',\n",
       " 'Reactome_2016',\n",
       " 'Reactome_2022',\n",
       " 'SILAC_Phosphoproteomics',\n",
       " 'SubCell_BarCode',\n",
       " 'SynGO_2022',\n",
       " 'SysMyo_Muscle_Gene_Sets',\n",
       " 'TF-LOF_Expression_from_GEO',\n",
       " 'TF_Perturbations_Followed_by_Expression',\n",
       " 'TG_GATES_2020',\n",
       " 'TRANSFAC_and_JASPAR_PWMs',\n",
       " 'TRRUST_Transcription_Factors_2019',\n",
       " 'Table_Mining_of_CRISPR_Studies',\n",
       " 'Tabula_Muris',\n",
       " 'Tabula_Sapiens',\n",
       " 'TargetScan_microRNA',\n",
       " 'TargetScan_microRNA_2017',\n",
       " 'The_Kinase_Library_2023',\n",
       " 'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
       " 'Tissue_Protein_Expression_from_ProteomicsDB',\n",
       " 'Transcription_Factor_PPIs',\n",
       " 'UK_Biobank_GWAS_v1',\n",
       " 'Virus-Host_PPI_P-HIPSTer_2020',\n",
       " 'VirusMINT',\n",
       " 'Virus_Perturbations_from_GEO_down',\n",
       " 'Virus_Perturbations_from_GEO_up',\n",
       " 'WikiPathway_2021_Human',\n",
       " 'WikiPathways_2013',\n",
       " 'WikiPathways_2015',\n",
       " 'WikiPathways_2016',\n",
       " 'WikiPathways_2019_Human',\n",
       " 'WikiPathways_2019_Mouse',\n",
       " 'dbGaP',\n",
       " 'huMAP',\n",
       " 'lncHUB_lncRNA_Co-Expression',\n",
       " 'miRTarBase_2017']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp.get_library_name()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5f665339",
   "metadata": {},
   "outputs": [],
   "source": [
    "pre_res = gp.prerank(rnk = ranking, gene_sets = 'GO_Biological_Process_2021', seed = 6, permutation_num = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "60a0dc53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>fdr</th>\n",
       "      <th>es</th>\n",
       "      <th>nes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNA replication (GO:0006260)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.726485</td>\n",
       "      <td>-2.307529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mitotic spindle organization (GO:0007052)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.741222</td>\n",
       "      <td>-2.291692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>microtubule cytoskeleton organization involved...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.755046</td>\n",
       "      <td>-2.272015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNA metabolic process (GO:0006259)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.662867</td>\n",
       "      <td>-2.236690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNA-dependent DNA replication (GO:0006261)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.715728</td>\n",
       "      <td>-2.178290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>positive regulation of translation (GO:0045727)</td>\n",
       "      <td>0.999371</td>\n",
       "      <td>-0.144227</td>\n",
       "      <td>-0.367706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>protein import (GO:0017038)</td>\n",
       "      <td>0.999574</td>\n",
       "      <td>0.121368</td>\n",
       "      <td>0.347975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>regulation of translation (GO:0006417)</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.136522</td>\n",
       "      <td>0.433231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>nuclear-transcribed mRNA catabolic process (GO...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.187822</td>\n",
       "      <td>0.583059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>regulation of ATP metabolic process (GO:1903578)</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.198150</td>\n",
       "      <td>0.597092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  Term       fdr        es  \\\n",
       "0                         DNA replication (GO:0006260)  0.000000 -0.726485   \n",
       "1            mitotic spindle organization (GO:0007052)  0.000000 -0.741222   \n",
       "2    microtubule cytoskeleton organization involved...  0.000000 -0.755046   \n",
       "3                   DNA metabolic process (GO:0006259)  0.000000 -0.662867   \n",
       "4           DNA-dependent DNA replication (GO:0006261)  0.000000 -0.715728   \n",
       "..                                                 ...       ...       ...   \n",
       "185    positive regulation of translation (GO:0045727)  0.999371 -0.144227   \n",
       "186                        protein import (GO:0017038)  0.999574  0.121368   \n",
       "187             regulation of translation (GO:0006417)  1.000000  0.136522   \n",
       "188  nuclear-transcribed mRNA catabolic process (GO...  1.000000  0.187822   \n",
       "189   regulation of ATP metabolic process (GO:1903578)  1.000000  0.198150   \n",
       "\n",
       "          nes  \n",
       "0   -2.307529  \n",
       "1   -2.291692  \n",
       "2   -2.272015  \n",
       "3   -2.236690  \n",
       "4   -2.178290  \n",
       "..        ...  \n",
       "185 -0.367706  \n",
       "186  0.347975  \n",
       "187  0.433231  \n",
       "188  0.583059  \n",
       "189  0.597092  \n",
       "\n",
       "[190 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out = []\n",
    "\n",
    "for term in list(pre_res.results):\n",
    "    out.append([term,\n",
    "               pre_res.results[term]['fdr'],\n",
    "               pre_res.results[term]['es'],\n",
    "               pre_res.results[term]['nes']])\n",
    "\n",
    "out_df = pd.DataFrame(out, columns = ['Term','fdr', 'es', 'nes']).sort_values(['fdr', 'nes']).reset_index(drop = True)\n",
    "out_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "03d7503c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'microtubule cytoskeleton organization involved in mitosis (GO:1902850)'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "term_to_graph = out_df.iloc[2].Term\n",
    "term_to_graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6ed09187",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Red and Blue means up-regulated and down-regulated genes. The green verticle bar means there is a hit from the gene set 'GO:1902850'\n",
    "# NES is net enrichment score and it means the geneset is negatively enriched\n",
    "gseaplot(rank_metric= pre_res.ranking, term = term_to_graph, **pre_res.results[term_to_graph],\n",
    "          ofname='gsea.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "968fd462",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f72cc918",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67a1e41a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f475465",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "810dde0f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ad17877",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79b7bc3a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0acdecdc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38cd40d9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40034503",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Example of user set of genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0591bf0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this block is just me making a enriched set of genes in the DE upregulated genes\n",
    "#not required for GSEA\n",
    "import random\n",
    "\n",
    "def a_thing(x):\n",
    "    if x < 0:\n",
    "        return 1\n",
    "    else:\n",
    "        return x\n",
    "df['weights'] = df.Rank.map(a_thing)\n",
    "\n",
    "a_list = random.choices(df.Gene.values, k = 100, weights = df.weights.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa1047d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "user_set = {'set_name1' : a_list}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e34daff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pre_res = gp.prerank(rnk = ranking, gene_sets = user_set, seed = 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b06c517",
   "metadata": {},
   "outputs": [],
   "source": [
    "gseaplot(pre_res.ranking, term = 'set_name1', **pre_res.results['set_name1']) #ofname = 'saved_fig.png'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43cd3dd6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
